README.md
# Flutter Paginator

[![pub package](https://img.shields.io/badge/pub-v2.0.0-orange)](https://pub.dev/packages/flutter_paginator_ns)

A customizable pagination package for Flutter. This package can be used 
to fetch and display data from an API page by page.

## NULL-SAFETY SUPPORT!!!
This package has been migrated to `null-safety`
- Warning - This package might be removed if the changes gets approved on the original package.

[Original Package](https://pub.dev/packages/flutter_paginator)


## Screenshots

<img src="https://raw.githubusercontent.com/bernardoveras/FlutterPaginator/master/screenrecord.gif" height="640em"/>

## Installing

```
dependencies:
  flutter_paginator_ns: ^2.0.0
```

## Description

`Paginator` extends `StatefulWidget` has 3 constructors namely `Paginator.listView`, `Paginator.gridView` and `Paginator.pageView`. `Paginator.listView`, `Paginator.gridView` and `Paginator.pageView` are descendants of `ListView`, `GridView` and `PageView`. `Paginator.listView`, `Paginator.gridView` and `Paginator.pageView` got all the features of their ancestors and they are need to provide additional functions that are essential in doing their task.

### `PageLoadFuture`
 * Loads the page asynchronously when the page number is given.
 * This should return an instance of a `Future`.
 * Called when the next page is needed to be loaded.

### `PageItemsGetter`
 * This function should return list of page item data when page data is given.
 * This is called after successful completion of `PageLoadFuture`.
 * The page items returned by this method is added to the list of all the
    page items.

### `ListItemBuilder`
 * Builds list item when item data and item index are given.
 * This should return an instance of a `Widget`.

### `LoadingWidgetBuilder`
 * Builds loading widget.
 * This should return an instance of a `Widget`.

### `ErrorWidgetBuilder`
 * Builds error widget when page data and error callback are given.
 * This should return an instance of a `Widget`.

### `EmptyListWidgetBuilder`
 * Builds empty list widget.
 * This is displayed when the total number of list items is zero.
 * This should return an instance of a `Widget`.

### `TotalItemsGetter`
 * This should return total number of list items when page data is given.

### `PageErrorChecker`
 * This should return true if page has error else false, when page data is given.
 
## Using

```dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

import 'package:flutter_paginator_ns/flutter_paginator.dart';
import 'package:flutter_paginator_ns/enums.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Paginator',
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return HomeState();
  }
}

class HomeState extends State<HomePage> {
  GlobalKey<PaginatorState> paginatorGlobalKey = GlobalKey();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Flutter Paginator'),
        actions: <Widget>[
          IconButton(
            icon: Icon(Icons.format_list_bulleted),
            onPressed: () {
              paginatorGlobalKey.currentState
                  .changeState(listType: ListType.LIST_VIEW);
            },
          ),
          IconButton(
            icon: Icon(Icons.grid_on),
            onPressed: () {
              paginatorGlobalKey.currentState.changeState(
                listType: ListType.GRID_VIEW,
                gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                    crossAxisCount: 2),
              );
            },
          ),
          IconButton(
            icon: Icon(Icons.library_books),
            onPressed: () {
              paginatorGlobalKey.currentState
                  .changeState(listType: ListType.PAGE_VIEW);
            },
          ),
        ],
      ),
      body: Paginator.listView(
        key: paginatorGlobalKey,
        pageLoadFuture: sendCountriesDataRequest,
        pageItemsGetter: listItemsGetter,
        listItemBuilder: listItemBuilder,
        loadingWidgetBuilder: loadingWidgetMaker,
        errorWidgetBuilder: errorWidgetMaker,
        emptyListWidgetBuilder: emptyListWidgetMaker,
        totalItemsGetter: totalPagesGetter,
        pageErrorChecker: pageErrorChecker,
        scrollPhysics: BouncingScrollPhysics(),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          paginatorGlobalKey.currentState.changeState(
              pageLoadFuture: sendCountriesDataRequest, resetState: true);
        },
        child: Icon(Icons.refresh),
      ),
    );
  }

  Future<CountriesData> sendCountriesDataRequest(int page) async {
    try {
      String url = Uri.encodeFull(
          'http://api.worldbank.org/v2/country?page=$page&format=json');
      http.Response response = await http.get(url);
      return CountriesData.fromResponse(response);
    } catch (e) {
      if (e is IOException) {
        return CountriesData.withError(
            'Please check your internet connection.');
      } else {
        print(e.toString());
        return CountriesData.withError('Something went wrong.');
      }
    }
  }

  List<dynamic> listItemsGetter(CountriesData countriesData) {
    List<String> list = [];
    countriesData.countries.forEach((value) {
      list.add(value['name']);
    });
    return list;
  }

  Widget listItemBuilder(value, int index) {
    return ListTile(
      leading: Text(index.toString()),
      title: Text(value),
    );
  }

  Widget loadingWidgetMaker() {
    return Container(
      alignment: Alignment.center,
      height: 160.0,
      child: CircularProgressIndicator(),
    );
  }

  Widget errorWidgetMaker(CountriesData countriesData, retryListener) {
    return Column(
      mainAxisAlignment: MainAxisAlignment.center,
      children: <Widget>[
        Padding(
          padding: const EdgeInsets.all(16.0),
          child: Text(countriesData.errorMessage),
        ),
        FlatButton(
          onPressed: retryListener,
          child: Text('Retry'),
        )
      ],
    );
  }

  Widget emptyListWidgetMaker(CountriesData countriesData) {
    return Center(
      child: Text('No countries in the list'),
    );
  }

  int totalPagesGetter(CountriesData countriesData) {
    return countriesData.total;
  }

  bool pageErrorChecker(CountriesData countriesData) {
    return countriesData.statusCode != 200;
  }
}

class CountriesData {
  List<dynamic> countries;
  int statusCode;
  String errorMessage;
  int total;
  int nItems;

  CountriesData.fromResponse(http.Response response) {
    this.statusCode = response.statusCode;
    List jsonData = json.decode(response.body);
    countries = jsonData[1];
    total = jsonData[0]['total'];
    nItems = countries.length;
  }

  CountriesData.withError(String errorMessage) {
    this.errorMessage = errorMessage;
  }
}
```

CHANGELOG.md
## [2.0.0] - NULL-SAFETY!

* Package migrated to null-safety

## [0.0.6] - Release

* A customizable paginator library written in dart.

flutter_paginator_example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_paginator_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/flutter_paginator.dart
library flutter_paginator_ns;

import 'dart:async';

import 'package:flutter/material.dart' hide ErrorWidgetBuilder;

import 'enums.dart';
import 'type_definitions.dart';

/// Paginator has 3 constructors
/// [Paginator.listView], [Paginator.gridView], [Paginator.pageView]
/// [Paginator.listView], [Paginator.gridView] and [Paginator.pageView] are
/// descendants of [ListView], [GridView] and [PageView].
/// [Paginator.listView], [Paginator.gridView] and [Paginator.pageView] got
/// all the features of their ancestors and they are need to provide additional
/// properties that are essential in doing their task.
///
/// [PageLoadFuture]
///  * Loads the page asynchronously when the page number is given.
///  * This should return an instance of a [Future].
///  * Called when the next page is needed to be loaded.
///
/// [PageItemsGetter]
///  * This function should return list of page item data when page data is given.
///  * This is called after successful completion of [pageLoadFuture].
///  * The page items returned by this method is added to the list of all the
///     page items.
///
/// [ListItemBuilder]
///  * Builds list item when item data and item index are given.
///  * This should return an instance of a [Widget].
///
/// [LoadingWidgetBuilder]
///  * Builds loading widget.
///  * This should return an instance of a [Widget].
///
/// [ErrorWidgetBuilder]
///  * Builds error widget when page data and error callback are given.
///  * This should return an instance of a [Widget].
///
/// [EmptyListWidgetBuilder]
///  * Builds empty list widget.
///  * This is displayed when the total number of list items is zero.
///  * This should return an instance of a [Widget].
///
/// [TotalItemsGetter]
///  * This should return total number of list items when page data is given.
///
/// [PageErrorChecker]
///  * This should return true if page has error else false, when page data is given.

class Paginator<T> extends StatefulWidget {
  final PageLoadFuture<T> pageLoadFuture;
  final PageItemsGetter<T> pageItemsGetter;
  final ListItemBuilder listItemBuilder;
  final LoadingWidgetBuilder loadingWidgetBuilder;
  final ErrorWidgetBuilder<T> errorWidgetBuilder;
  final EmptyListWidgetBuilder<T> emptyListWidgetBuilder;
  final TotalItemsGetter<T> totalItemsGetter;
  final PageErrorChecker<T> pageErrorChecker;

  /// common properties
  final Key? scrollViewKey;
  final ScrollPhysics? scrollPhysics;
  final Axis scrollDirection;
  final bool reverse;

  /// properties - list view, grid view
  final EdgeInsets? padding;
  final bool shrinkWrap;
  final ScrollController? scrollController;
  final bool? primary;
  final int? semanticChildCount;
  final double? cacheExtent;
  final bool addAutomaticKeepAlives;
  final bool addRepaintBoundaries;
  final bool addSemanticIndexes;

  /// properties - list view only
  final double? itemExtent;

  /// properties - grid view only
  final SliverGridDelegate? gridDelegate;

  /// properties - page view only
  final ValueChanged<int>? onPageChanged;
  final bool pageSnapping;
  final PageController? pageController;

  final ListType listType;

  const Paginator.listView({
    Key? key,
    required this.pageLoadFuture,
    required this.pageItemsGetter,
    required this.listItemBuilder,
    required this.loadingWidgetBuilder,
    required this.errorWidgetBuilder,
    required this.emptyListWidgetBuilder,
    required this.totalItemsGetter,
    required this.pageErrorChecker,
    this.scrollViewKey,
    this.scrollDirection = Axis.vertical,
    this.reverse = false,
    this.scrollPhysics,
    this.padding,
    this.shrinkWrap = false,
    this.scrollController,
    this.primary,
    this.semanticChildCount,
    this.cacheExtent,
    this.addAutomaticKeepAlives = true,
    this.addRepaintBoundaries = true,
    this.addSemanticIndexes = true,
    this.itemExtent,
  })  : this.listType = ListType.LIST_VIEW,
        this.onPageChanged = null,
        this.pageSnapping = true,
        this.pageController = null,
        this.gridDelegate = null,
        super(key: key);

  const Paginator.gridView({
    Key? key,
    required this.pageLoadFuture,
    required this.pageItemsGetter,
    required this.listItemBuilder,
    required this.loadingWidgetBuilder,
    required this.errorWidgetBuilder,
    required this.emptyListWidgetBuilder,
    required this.totalItemsGetter,
    required this.pageErrorChecker,
    required this.gridDelegate,
    this.scrollViewKey,
    this.scrollDirection = Axis.vertical,
    this.reverse = false,
    this.scrollPhysics,
    this.padding,
    this.shrinkWrap = false,
    this.scrollController,
    this.primary,
    this.semanticChildCount,
    this.cacheExtent,
    this.addAutomaticKeepAlives = true,
    this.addRepaintBoundaries = true,
    this.addSemanticIndexes = true,
  })  : this.listType = ListType.GRID_VIEW,
        this.itemExtent = null,
        this.onPageChanged = null,
        this.pageSnapping = true,
        this.pageController = null,
        super(key: key);

  const Paginator.pageView({
    Key? key,
    required this.pageLoadFuture,
    required this.pageItemsGetter,
    required this.listItemBuilder,
    required this.loadingWidgetBuilder,
    required this.errorWidgetBuilder,
    required this.emptyListWidgetBuilder,
    required this.totalItemsGetter,
    required this.pageErrorChecker,
    this.scrollViewKey,
    this.scrollDirection = Axis.horizontal,
    this.reverse = false,
    this.scrollPhysics,
    this.onPageChanged,
    this.pageSnapping = true,
    this.pageController,
  })  : this.listType = ListType.PAGE_VIEW,
        this.padding = null,
        this.shrinkWrap = false,
        this.scrollController = null,
        this.primary = null,
        this.semanticChildCount = null,
        this.cacheExtent = null,
        this.addAutomaticKeepAlives = true,
        this.addRepaintBoundaries = true,
        this.addSemanticIndexes = true,
        this.itemExtent = null,
        this.gridDelegate = null,
        super(key: key);

  @override
  State<StatefulWidget> createState() {
    return PaginatorState<T>(
      this.pageLoadFuture,
      this.pageItemsGetter,
      this.listItemBuilder,
      this.loadingWidgetBuilder,
      this.errorWidgetBuilder,
      this.emptyListWidgetBuilder,
      this.totalItemsGetter,
      this.pageErrorChecker,
      this.gridDelegate,
      this.scrollViewKey,
      this.scrollDirection,
      this.reverse,
      this.scrollPhysics,
      this.padding,
      this.shrinkWrap,
      this.scrollController,
      this.primary,
      this.semanticChildCount,
      this.cacheExtent,
      this.addAutomaticKeepAlives,
      this.addRepaintBoundaries,
      this.addSemanticIndexes,
      this.itemExtent,
      this.onPageChanged,
      this.pageSnapping,
      this.pageController,
      this.listType,
    );
  }
}

class PaginatorState<T> extends State<Paginator> {
  /// essential
  PageLoadFuture<T> _pageLoadFuture;
  PageItemsGetter<T> _pageItemsGetter;
  ListItemBuilder _listItemBuilder;
  LoadingWidgetBuilder _loadingWidgetBuilder;
  ErrorWidgetBuilder<T> _errorWidgetBuilder;
  EmptyListWidgetBuilder<T> _emptyListWidgetBuilder;
  TotalItemsGetter<T> _totalItemsGetter;
  PageErrorChecker<T> _pageErrorChecker;

  /// common properties
  Key? _scrollViewKey;
  ScrollPhysics? _scrollPhysics;
  Axis _scrollDirection;
  bool _reverse;

  /// properties - list view, grid view
  EdgeInsets? _padding;
  bool _shrinkWrap;
  ScrollController? _scrollController;
  bool? _primary;
  int? _semanticChildCount;
  double? _cacheExtent;
  bool _addAutomaticKeepAlives;
  bool _addRepaintBoundaries;
  bool _addSemanticIndexes;

  /// properties - list view only
  double? _itemExtent;

  /// properties - grid view only
  SliverGridDelegate? _gridDelegate;

  /// properties - page view only
  ValueChanged<int>? _onPageChanged;
  bool _pageSnapping;
  PageController? _pageController;

  /// runtime variables
  late _ListStatus _listStatus;
  ListType _listType;
  late List _listItems;
  late int _currentPage;
  late int _nTotalItems;
  late T _firstPageData;

  PaginatorState(
    this._pageLoadFuture,
    this._pageItemsGetter,
    this._listItemBuilder,
    this._loadingWidgetBuilder,
    this._errorWidgetBuilder,
    this._emptyListWidgetBuilder,
    this._totalItemsGetter,
    this._pageErrorChecker,
    this._gridDelegate,
    this._scrollViewKey,
    this._scrollDirection,
    this._reverse,
    this._scrollPhysics,
    this._padding,
    this._shrinkWrap,
    this._scrollController,
    this._primary,
    this._semanticChildCount,
    this._cacheExtent,
    this._addAutomaticKeepAlives,
    this._addRepaintBoundaries,
    this._addSemanticIndexes,
    this._itemExtent,
    this._onPageChanged,
    this._pageSnapping,
    this._pageController,
    this._listType,
  );

  @override
  void initState() {
    _initialize();
    _initialFutureCall();
    super.initState();
  }

  void _initialize() {
    _listStatus = _ListStatus.LOADING;
    _listItems = [];
    _currentPage = 0;
    _nTotalItems = 0;
  }

  @override
  Widget build(BuildContext context) {
    switch (_listStatus) {
      case _ListStatus.SUCCESS:
        return _buildPaginatorWidget();

      case _ListStatus.LOADING:
        return Center(child: _loadingWidgetBuilder());

      case _ListStatus.ERROR:
        return Center(child: _errorWidgetBuilder(_firstPageData, _onError));

      case _ListStatus.EMPTY:
      default:
        return _emptyListWidgetBuilder(_firstPageData);
    }
  }

  Widget _buildPaginatorWidget() {
    switch (_listType) {
      case ListType.LIST_VIEW:
        return ListView.builder(
          key: _scrollViewKey,
          padding: _padding,
          physics: _scrollPhysics,
          scrollDirection: _scrollDirection,
          shrinkWrap: _shrinkWrap,
          itemCount: _getItemCount(),
          itemBuilder: _itemBuilder,
          controller: _scrollController,
          reverse: _reverse,
          primary: _primary,
          semanticChildCount: _semanticChildCount,
          cacheExtent: _cacheExtent,
          itemExtent: _itemExtent,
          addAutomaticKeepAlives: _addAutomaticKeepAlives,
          addRepaintBoundaries: _addRepaintBoundaries,
          addSemanticIndexes: _addSemanticIndexes,
        );
      case ListType.GRID_VIEW:
        return GridView.builder(
          key: _scrollViewKey,
          gridDelegate: _gridDelegate!,
          padding: _padding,
          physics: _scrollPhysics,
          scrollDirection: _scrollDirection,
          shrinkWrap: _shrinkWrap,
          itemCount: _getItemCount(),
          itemBuilder: _itemBuilder,
          controller: _scrollController,
          reverse: _reverse,
          primary: _primary,
          semanticChildCount: _semanticChildCount,
          cacheExtent: _cacheExtent,
          addAutomaticKeepAlives: _addAutomaticKeepAlives,
          addRepaintBoundaries: _addRepaintBoundaries,
          addSemanticIndexes: _addSemanticIndexes,
        );
      case ListType.PAGE_VIEW:
      default:
        return PageView.builder(
          key: _scrollViewKey,
          physics: _scrollPhysics,
          scrollDirection: _scrollDirection,
          itemCount: _getItemCount(),
          itemBuilder: _itemBuilder,
          controller: _pageController,
          reverse: _reverse,
          onPageChanged: _onPageChanged,
          pageSnapping: _pageSnapping,
        );
    }
  }

  Widget _itemBuilder(BuildContext context, int index) {
    if (index < _listItems.length) {
      return _listItemBuilder(_listItems[index], index);
    } else {
      return FutureBuilder(
        future: _pageLoadFuture(_currentPage + 1),
        builder: (BuildContext context, AsyncSnapshot<T> snapshot) {
          switch (snapshot.connectionState) {
            case ConnectionState.done:
              if (_pageErrorChecker(snapshot.data!)) {
                return _errorWidgetBuilder(snapshot.data!, _onError);
              }
              _listItems.addAll(_pageItemsGetter(snapshot.data!));
              _currentPage++;
              Future.microtask(() {
                setState(() {});
              });
              return _loadingWidgetBuilder();
            default:
              return _loadingWidgetBuilder();
          }
        },
      );
    }
  }

  int? _getItemCount() {
    if (_nTotalItems > _listItems.length) {
      return _listItems.length + 1;
    }
    return _nTotalItems;
  }

  void _onError() {
    if (_listItems.length == 0) {
      setState(
        () {
          this._listStatus = _ListStatus.LOADING;
        },
      );
      _initialFutureCall();
    } else {
      setState(() {});
    }
  }

  void changeState({
    PageLoadFuture<T>? pageLoadFuture,
    PageItemsGetter<T?>? pageItemsGetter,
    ListItemBuilder? listItemBuilder,
    LoadingWidgetBuilder? loadingWidgetBuilder,
    ErrorWidgetBuilder<T?>? errorWidgetBuilder,
    EmptyListWidgetBuilder<T?>? emptyListWidgetMaker,
    TotalItemsGetter<T>? totalItemsGetter,
    PageErrorChecker<T?>? pageErrorChecker,
    Key? scrollViewKey,
    ScrollPhysics? scrollPhysics,
    Axis? scrollDirection,
    bool? reverse,
    EdgeInsets? padding,
    bool? shrinkWrap,
    ScrollController? scrollController,
    bool? primary,
    int? semanticChildCount,
    double? cacheExtent,
    bool? addAutomaticKeepAlives,
    bool? addRepaintBoundaries,
    bool? addSemanticIndexes,
    double? itemExtent,
    SliverGridDelegate? gridDelegate,
    ValueChanged<int>? onPageChanged,
    bool? pageSnapping,
    PageController? pageController,
    ListType? listType,
    bool resetState = false,
  }) {
    _pageLoadFuture = pageLoadFuture ?? _pageLoadFuture;
    _pageItemsGetter = pageItemsGetter ?? _pageItemsGetter;
    _listItemBuilder = listItemBuilder ?? _listItemBuilder;
    _loadingWidgetBuilder = loadingWidgetBuilder ?? _loadingWidgetBuilder;
    _errorWidgetBuilder = errorWidgetBuilder ?? _errorWidgetBuilder;
    _emptyListWidgetBuilder = emptyListWidgetMaker ?? _emptyListWidgetBuilder;
    _totalItemsGetter = totalItemsGetter ?? _totalItemsGetter;
    _pageErrorChecker = pageErrorChecker ?? _pageErrorChecker;
    _scrollViewKey = scrollViewKey ?? _scrollViewKey;
    _scrollPhysics = scrollPhysics ?? _scrollPhysics;
    _scrollDirection = scrollDirection ?? _scrollDirection;
    _reverse = reverse ?? _reverse;
    _padding = padding ?? _padding;
    _shrinkWrap = shrinkWrap ?? _shrinkWrap;
    _scrollController = scrollController ?? _scrollController;
    _primary = primary ?? _primary;
    _semanticChildCount = semanticChildCount ?? _semanticChildCount;
    _cacheExtent = cacheExtent ?? _cacheExtent;
    _addAutomaticKeepAlives = addAutomaticKeepAlives ?? _addAutomaticKeepAlives;
    _addRepaintBoundaries = addRepaintBoundaries ?? _addRepaintBoundaries;
    _addSemanticIndexes = addSemanticIndexes ?? _addSemanticIndexes;
    _itemExtent = itemExtent ?? _itemExtent;
    _gridDelegate = gridDelegate ?? _gridDelegate;
    _onPageChanged = onPageChanged ?? _onPageChanged;
    _pageSnapping = pageSnapping ?? _pageSnapping;
    _pageController = pageController ?? _pageController;
    _listType = listType ?? _listType;

    if (resetState) {
      _initialize();
      setState(() {});
      _initialFutureCall();
    }
    setState(() {});
  }

  void _initialFutureCall() {
    Future<T> future = _pageLoadFuture(1);
    future.then((T pageData) {
      _firstPageData = pageData;
      _nTotalItems = _totalItemsGetter(pageData);
      if (_pageErrorChecker(pageData)) {
        setState(() {
          _listStatus = _ListStatus.ERROR;
        });
      } else if (_nTotalItems == 0) {
        setState(() {
          _listStatus = _ListStatus.EMPTY;
        });
      } else {
        _currentPage++;
        _listItems.addAll(_pageItemsGetter(pageData));
        setState(() {
          _listStatus = _ListStatus.SUCCESS;
        });
      }
    });
  }
}

enum _ListStatus { LOADING, ERROR, SUCCESS, EMPTY }


example/lib/main.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

import 'package:flutter_paginator_ns/flutter_paginator.dart';
import 'package:flutter_paginator_ns/enums.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Paginator',
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return HomeState();
  }
}

class HomeState extends State<HomePage> {
  final GlobalKey<PaginatorState> paginatorGlobalKey = GlobalKey();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Flutter Paginator'),
        actions: <Widget>[
          IconButton(
            icon: Icon(Icons.format_list_bulleted),
            onPressed: () {
              paginatorGlobalKey.currentState!.changeState(listType: ListType.LIST_VIEW);
            },
          ),
          IconButton(
            icon: Icon(Icons.grid_on),
            onPressed: () {
              paginatorGlobalKey.currentState!.changeState(
                listType: ListType.GRID_VIEW,
                gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(crossAxisCount: 2),
              );
            },
          ),
          IconButton(
            icon: Icon(Icons.library_books),
            onPressed: () {
              paginatorGlobalKey.currentState!.changeState(listType: ListType.PAGE_VIEW);
            },
          ),
        ],
      ),
      body: Paginator.listView(
        key: paginatorGlobalKey,
        pageLoadFuture: sendCountriesDataRequest,
        pageItemsGetter: listItemsGetter,
        listItemBuilder: listItemBuilder,
        loadingWidgetBuilder: loadingWidgetMaker,
        errorWidgetBuilder: errorWidgetMaker,
        emptyListWidgetBuilder: emptyListWidgetMaker,
        totalItemsGetter: totalPagesGetter,
        pageErrorChecker: pageErrorChecker,
        scrollPhysics: BouncingScrollPhysics(),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          paginatorGlobalKey.currentState!.changeState(pageLoadFuture: sendCountriesDataRequest, resetState: true);
        },
        child: Icon(Icons.refresh),
      ),
    );
  }

  Future<CountriesData> sendCountriesDataRequest(int page) async {
    try {
      Uri url = Uri.parse('http://api.worldbank.org/v2/country?page=$page&format=json');
      http.Response response = await http.get(url);
      return CountriesData.fromResponse(response);
    } catch (e) {
      if (e is IOException) {
        return CountriesData.withError('Please check your internet connection.');
      } else {
        print(e.toString());
        return CountriesData.withError('Something went wrong.');
      }
    }
  }

  List<dynamic> listItemsGetter(CountriesData countriesData) {
    List<String> list = [];
    countriesData.countries.forEach((value) {
      list.add(value['name']);
    });
    return list;
  }

  Widget listItemBuilder(value, int index) {
    return ListTile(
      leading: Text(index.toString()),
      title: Text(value),
    );
  }

  Widget loadingWidgetMaker() {
    return Container(
      alignment: Alignment.center,
      height: 160.0,
      child: CircularProgressIndicator(),
    );
  }

  Widget errorWidgetMaker(CountriesData countriesData, retryListener) {
    return Column(
      mainAxisAlignment: MainAxisAlignment.center,
      children: <Widget>[
        Padding(
          padding: const EdgeInsets.all(16.0),
          child: Text(countriesData.errorMessage),
        ),
        ElevatedButton(
          onPressed: retryListener,
          child: Text('Retry'),
        )
      ],
    );
  }

  Widget emptyListWidgetMaker(CountriesData countriesData) {
    return Center(
      child: Text('No countries in the list'),
    );
  }

  int totalPagesGetter(CountriesData countriesData) {
    return countriesData.total;
  }

  bool pageErrorChecker(CountriesData countriesData) {
    return countriesData.statusCode != 200;
  }
}

class CountriesData {
  List<dynamic> countries = <dynamic>[];
  late int statusCode;
  String errorMessage = '';
  int total = 0;
  int nItems = 0;

  CountriesData.fromResponse(http.Response response) {
    this.statusCode = response.statusCode;
    List jsonData = json.decode(response.body);
    countries = jsonData[1];
    total = jsonData[0]['total'];
    nItems = countries.length;
  }

  CountriesData.withError(String errorMessage) {
    this.errorMessage = errorMessage;
  }
}
